// @loopholelabs/polyglot/kind.js
"use strict";
/*
    Copyright 2023 Loophole Labs

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

           http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Kind = void 0;
var Kind;
(function (Kind) {
    Kind[Kind["Null"] = 0] = "Null";
    Kind[Kind["Array"] = 1] = "Array";
    Kind[Kind["Map"] = 2] = "Map";
    Kind[Kind["Any"] = 3] = "Any";
    Kind[Kind["Uint8Array"] = 4] = "Uint8Array";
    Kind[Kind["String"] = 5] = "String";
    Kind[Kind["Error"] = 6] = "Error";
    Kind[Kind["Boolean"] = 7] = "Boolean";
    Kind[Kind["Uint8"] = 8] = "Uint8";
    Kind[Kind["Uint16"] = 9] = "Uint16";
    Kind[Kind["Uint32"] = 10] = "Uint32";
    Kind[Kind["Uint64"] = 11] = "Uint64";
    Kind[Kind["Int32"] = 12] = "Int32";
    Kind[Kind["Int64"] = 13] = "Int64";
    Kind[Kind["Float32"] = 14] = "Float32";
    Kind[Kind["Float64"] = 15] = "Float64";
})(Kind || (exports.Kind = Kind = {}));
